### PR TITLE
hotfix find_within

### LINKED ISSUE #523: `find_within` bug with `distance_measure='spherical_radians'`
## General

* pyfar version: 0.6.1
* Python version: 3.9.1
* Operating System: MacOS
* Did you install pyfar via pip: yes

## Description

`find_within` bug with `distance_measure='spherical_radians'`
 
## What I Did

I want to use pf to find all point on the northern hemisphere, I used scripts below. It returns a crash.

```
find = pf.Coordinates(0, 0, 1)
spatial_mask = coords.find_within(
    find,
    distance=np.pi/2,
    distance_measure='spherical_radians',
    atol=0.1)
coords.show(spatial_mask[0])
```

I located bug here, change `==` to `in` help me solve the problem.

```
        if distance_measure == ['spherical_radians', 'spherical_meter']:
            # determine validate radius
            radius = self.radius
            delta_radius = np.max(radius) - np.min(radius)
            if delta_radius > 1e-15:
                raise ValueError(
                    "find_nearest_sph only works if all points have the same \
                    radius. Differences are larger than 1e-15")
            radius = np.max(radius)

```

furthermore, I'm interseted in the delta_radius, 1e-15 is very small so that float value is hard to satify this threshold. Could we consider lowering the threshold?

### PR DESCRIPTION
**Pull request by ahms5**
### Which issue(s) are closed by this pull request?

Closes #523

### Changes proposed in this pull request:

- fix bug and replace ``==`` to ``in``
- add test for spherical distance measure
- add doc and test for atol
- add new option ``radius_tol`` to change the tolerance for radius + add test

### CODE REVIEW

File: pyfar/classes/coordinates.py [hunk 1]
```diff
@@ -1568,7 +1568,7 @@ def find_nearest(self, find, k=1, distance_measure='euclidean'):
 
     def find_within(
             self, find, distance=0., distance_measure='euclidean',
-            atol=1e-15, return_sorted=True):
+            atol=1e-15, radius_tol=1e-15, return_sorted=True):
```
[the remark this answers is not in the document]
Comment (mberz) [reply pyfar/classes/coordinates.py:1571]: I'll have another look later. There's no opportunity for me to get at a computer.

**Review by mberz — CHANGES_REQUESTED** [review]

File: pyfar/classes/coordinates.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (f-brinkmann) [reply pyfar/classes/coordinates.py:1571]: I would suggest to move the new parameter to the end.

[the remark this answers is not in the document]
Comment (f-brinkmann) [reply pyfar/classes/coordinates.py:1571]: One general question. If I remember correctly, we are using cartesian coordinates in the background to find data, even if providin spherical distance measures. Does this mean thatfor large values of `radius_tol`, we could find points that are acutally further away than specified by `distance` or not find points, that are closer then `distance``? If yes, I would suggest to mention it in the docstring.

**Review by f-brinkmann — CHANGES_REQUESTED** [review]
Thanks for implementing. I had one question regarding the behavior of the function.

File: pyfar/classes/coordinates.py [hunk 2]
```diff
@@ -1590,6 +1590,11 @@ def find_within(
             ``'spherical_meter'``
                 distance is determined by the great-circle distance
                 expressed in meters.
+        atol : float
+            Absolute tolerance for distance. The default is 1e-15.
+        radius_tol : float
+            Maximum allowed difference within radii for spherical distance
+            measures. The default is 1e-15.
```
Comment (pingelit) [inline pyfar/classes/coordinates.py:1597]: From this explanation, I don't understand what happens. Maybe a general explanation regarding tolerance could help. ... Or I#m just missing something.

**Review by pingelit — CHANGES_REQUESTED** [review]

File: pyfar/classes/coordinates.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ahms5) [reply pyfar/classes/coordinates.py:1571]: I added it in the docs

File: pyfar/classes/coordinates.py (hunk 2 above)
Comment (ahms5) [reply pyfar/classes/coordinates.py:1597]: fixed

File: pyfar/classes/coordinates.py [hunk 3]
```diff
@@ -1590,14 +1590,18 @@ def find_within(
...
                 distance is determined by the great-circle distance
                 expressed in meters.
-        atol : float
-            Absolute tolerance for distance. The default is 1e-15.
-        radius_tol : float
-            Maximum allowed difference within radii for spherical distance
-            measures. The default is 1e-15.
+        atol : float, None
+            Absolute tolerance for distance. The default is 2*eps, where eps
+            is the resolution of the data type of the coordinates.
```
Comment (f-brinkmann) [inline pyfar/classes/coordinates.py:1595]: If were not using eps, I would suggest to be a bit more clear in the docsring.
```suggestion
            Absolute tolerance for distance. The default ``None`` uses a tolerance of two times the decimal resolution, which is determined from the data type of the coordinate points using ``numpy.finfo``.
```

File: pyfar/classes/coordinates.py [hunk 4]
```diff
@@ -1590,14 +1590,18 @@ def find_within(
...
         return_sorted : bool, optional
             Sorts returned indices if True and does not sort them if False.
             The default is True.
+        radius_tol : float, None
+            For all spherical distance measures, the coordinates must be on
+            a sphere, so the radius must be constant. This parameter defines
+            the maximum allowed difference within the radii. Note that too
+            large tolerances can lead to wrong results. The default is 2*eps,
```
Comment (f-brinkmann) [inline pyfar/classes/coordinates.py:1603]: Maybe be more explicit and write:

Note that increasing the tolearance decreases the accuracy of the search, i.e., points that are within the search distance might not be found or points outside the search distance may be returned.

File: pyfar/classes/coordinates.py [hunk 5]
```diff
@@ -1590,14 +1590,18 @@ def find_within(
...
+            see ``atol`` for more details.
```
Comment (f-brinkmann) [inline pyfar/classes/coordinates.py:1604]: I would suggest to be more explicit as above and not refer to `atol`

**Review by f-brinkmann — CHANGES_REQUESTED** [review]
I made some suggestions that might improve the docstring and learned the difference between finfo.resolution and finfo.eps - thanks

**Review by pingelit — COMMENTED** [review]
Thanks for adapting the documentation. 

I agree with @f-brinkmann; being more specific would help even more. When it's changed I will approve.

**Review by pingelit — APPROVED** [review]
Thanks!

**Review by f-brinkmann — APPROVED** [review]

**Pull request merged**